bootloader/bootloader.go

```diff
@@ -252,7 +252,7 @@ func InstallBootConfig(gadgetDir, rootDir string, opts *Options) error {
 	return bl.InstallBootConfig(gadgetDir, opts)
 }
 
-type bootloaderNewFunc func(rootdir string, opts *Options) Bootloader
+type bootloaderNewFunc func(rootdir string, opts *Options) (Bootloader, error)
```

Comment:
I'm probably missing something but to me it doesn't make sense for ConfigFile and also the constructor to return errors ?  unless you need some kind of ConfigFile internal error ? basically I need to understand the use case more before approving the ConfigFile change

Reply:
Well for the new LK implementation, we don't know the static config file location at runtime, because it's not a static path, we have to query things to know which disk and which partition is the "config file". I.e. the bootloader sets something like `snapd_lk_boot_disk=mmcblk0`, so I have to parse `mmcblk0` from that, then go looking on `/dev/mmcblk0` and find the partition with label "snapbootsel", i.e. `/dev/mmcblk0p8`, and at that point then we can say the "config file" for LK is `/dev/mmcblk0p8`. 

The config file at image build time is always the same, but at runtime it is quite a bit trickier.

Reply:
that in itself sounds painful but fine. my point is unclear why both new and ConfigFile need to return errors,  and whether we can't just do the expensive work in new?  (I know that some details aren't right for the latter, but skimming quickly that seems fixable)

Reply:
I guess to be more clear, I really don't think we have the way of thinking about identifying bootloaders, it's very weird to say we only have a given bootloader on an installed core system when we can find a given "config file". I think we should really move away from this method of searching for any bootloader to instead just looking for the bootloader we know we should be using on that system - any ubuntu core system will have a gadget.yaml and the gadget.yaml has to say what the bootloader is, and so I think we should just do this in Find():

1. see if we are on Ubuntu Core, if not return err/not-found
1. if we are on Ubuntu Core we must have a gadget snap, go find it's gadget.yaml
1. get the bootloader from the gadget snap's gadget.yaml
1. try to create one of those bootloader types
1. if that step fails give up 

Then we can have things like `StaticGadgetConfigFile() string` which i.e. always returns the `bl.Name()+".conf"` thing that we need from various other places, without introducing this weirdness.

but all of that is a much larger change and I'm now suddenly worried that doing all this will significantly delay landing the feature given the widespread usage of the bootloader package across at least tests :-/